### PR TITLE
Disallow property assignment to complex data types

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Prior to this PR, we allowed properties to be assigned to complex types (i.e. nodes and edges). This is actually a fairly serious bug, as it causes crashes in RDB serialization.

### CODE REVIEW

File: src/graph/entities/graph_entity.c [hunk 1]
```diff
@@ -48,6 +49,14 @@ static void _GraphEntity_RemoveProperty(const GraphEntity *e, Attribute_ID attr_
 
 /* Add a new property to entity */
 SIValue *GraphEntity_AddProperty(GraphEntity *e, Attribute_ID attr_id, SIValue value) {
+	ASSERT(e);
+
+	// Emit an error and return if we're trying to add an invalid type.
+	if(!(SI_TYPE(value) & SI_PRIMITIVES)) {
+		QueryCtx_SetError("Property values can only be of primitive types.");
```
Comment (swilly22) [inline src/graph/entities/graph_entity.c:56]: Neo's error:
"Property values can only be of primitive types or arrays thereof"

File: src/graph/entities/graph_entity.c [hunk 2]
```diff
@@ -48,6 +49,14 @@ static void _GraphEntity_RemoveProperty(const GraphEntity *e, Attribute_ID attr_
 
 /* Add a new property to entity */
 SIValue *GraphEntity_AddProperty(GraphEntity *e, Attribute_ID attr_id, SIValue value) {
+	ASSERT(e);
+
+	// Emit an error and return if we're trying to add an invalid type.
+	if(!(SI_TYPE(value) & SI_PRIMITIVES)) {
```
Comment (swilly22) [inline src/graph/entities/graph_entity.c:55]: Catching the error at this point might be a bit too late
```
MATCH (l:L),(r:R),(q:Q) SET l.v = 1, r.v = 2, q.v = l
```
Might leave the graph in a none deterministic state, there isn't much we can do about it beside turn SET/UPDATE into an egger operation. I think the way to counter this is via a rollback functionality (MVCC)

File: src/graph/entities/graph_entity.c [hunk 3]
```diff
@@ -78,7 +87,13 @@ SIValue *GraphEntity_GetProperty(const GraphEntity *e, Attribute_ID attr_id) {
...
 // Updates existing property value.
 void GraphEntity_SetProperty(const GraphEntity *e, Attribute_ID attr_id, SIValue value) {
-	assert(e);
+	ASSERT(e);
+
+	// Emit an error and return if we're trying to add an invalid type.
+	if(!(SI_TYPE(value) & SI_PRIMITIVES)) {
+		QueryCtx_SetError("Property values can only be of primitive types.");
+		return;
+	}
```
Comment (swilly22) [inline src/graph/entities/graph_entity.c:96]: Same verification as in `GraphEntity_AddProperty` consider moving into a static inline function.

File: src/value.h [hunk 4]
```diff
@@ -50,6 +50,8 @@ typedef enum {
 #define SI_NUMERIC (T_INT64 | T_DOUBLE)
 #define SI_GRAPHENTITY (T_NODE | T_EDGE)
 #define SI_ALL (T_MAP | T_NODE | T_EDGE | T_ARRAY | T_PATH | T_DATETIME | T_LOCALDATETIME | T_DATE | T_TIME | T_LOCALTIME | T_DURATION | T_STRING | T_BOOL | T_INT64 | T_DOUBLE | T_NULL | T_PTR)
+#define SI_PRIMITIVES (T_ARRAY | T_DATETIME | T_LOCALDATETIME | T_DATE | T_TIME | T_LOCALTIME | T_DURATION | T_STRING | T_BOOL | T_INT64 | T_DOUBLE | T_NULL)
```
Comment (swilly22) [inline src/value.h:53]: Is ARRAY considered a primitive data type? In most languages arrays aren't primitives.

File: tests/flow/test_query_validation.py [hunk 5]
```diff
@@ -447,3 +447,22 @@ def test29_invalid_filter_non_boolean_constant(self):
         # Non-existent properties are treated as NULLs, which are boolean in Cypher's 3-valued logic.
         query = """MATCH (a) WHERE a.fakeprop RETURN a"""
         redis_graph.query(query)
+
+    def test31_set_invalid_property_type(self):
+        try:
+            query = """MATCH (a) CREATE (:L {v: a})"""
```
Comment (swilly22) [inline tests/flow/test_query_validation.py:453]: I guess there's a guarantee that the graph isn't empty.

File: tests/flow/test_query_validation.py [hunk 6]
```diff
@@ -447,3 +447,22 @@ def test29_invalid_filter_non_boolean_constant(self):
...
+            redis_graph.query(query)
+            assert(False)
+        except redis.exceptions.ResponseError as e:
+            # Expecting an error.
+            assert("Property values can only be of primitive types" in e.message)
+            pass
+
+        try:
+            query = """MATCH (a)-[]->(b) SET b.age = a"""
```
Comment (swilly22) [inline tests/flow/test_query_validation.py:462]: Make sure we're setting an existing attribute.
```
MATCH (a)-[]->(b) WHERE b.age = $age SET b.age = a
```

File: tests/flow/test_query_validation.py [hunk 7]
```diff
@@ -447,3 +447,22 @@ def test29_invalid_filter_non_boolean_constant(self):
         # Non-existent properties are treated as NULLs, which are boolean in Cypher's 3-valued logic.
         query = """MATCH (a) WHERE a.fakeprop RETURN a"""
         redis_graph.query(query)
+
+    def test31_set_invalid_property_type(self):
+        try:
```
Comment (swilly22) [inline tests/flow/test_query_validation.py:452]: if you want to make this test a bit more compact
```
queries = [q0, q1]
for q in queries:
    try:
        redis_graph.query(query)
        assert(False)
    except redis.exceptions.ResponseError as e:
        # Expecting an error.
        assert("Property values can only be of primitive types" in e.message)
        pass
```

**Review by swilly22 — CHANGES_REQUESTED** [review]
Clear and straight to the point, please see comments.

File: src/value.h (hunk 4 above)
Comment (jeffreylovitz) [reply src/value.h:53]: :+1: You're right, will fix.

File: src/graph/entities/graph_entity.c (hunk 2 above)
Comment (jeffreylovitz) [reply src/graph/entities/graph_entity.c]: You're right, I've added this logic to OpCreate and OpUpdate to catch errors before anything is committed. This still is a risk for MERGE...ON MATCH/CREATE SET, but the surface area is at least diminished.

**Review by swilly22 — APPROVED** [review]

**Pull request merged**